scripts/dts/python-devicetree/src/devicetree/edtlib.py

```diff
@@ -3116,21 +3115,39 @@ def _pass_thru(
     return res[-len(parent_spec):]
 
 
-def _raw_unit_addr(node: dtlib_Node) -> bytes:
+def _raw_unit_addr(node: dtlib_Node, parent: dtlib_Node) -> bytes:
     # _map_interrupt() helper. Returns the unit address (derived from 'reg' and
     # #address-cells) as a raw 'bytes'
 
+    iparent: Optional[dtlib_Node] = parent
+    addr_len: Optional[int] = None
+
+    while iparent is not None:
+        addr_len = _address_cells_self(iparent)
+        if addr_len is None:
+            # check the parent(s) of interrupt-controller(s) too
+            iparent = iparent.parent
+        else:
+            break
+
+    if addr_len is None:
+        addr_len =  2  # Default value per DT spec.
+
+    if addr_len == 0:
+        return b''
+
     if 'reg' not in node.props:
         _err(f"{node!r} lacks 'reg' property "
              "(needed for 'interrupt-map' unit address lookup)")
 
-    addr_len = 4*_address_cells(node)
-
-    if len(node.props['reg'].value) < addr_len:
-        _err(f"{node!r} has too short 'reg' property "
+    addr_len *= 4
+    prop_len = len(node.props['reg'].value)
+    if  prop_len < addr_len or prop_len % 4 != 0:
```

Comment:
Maybe also check that the `#address-cells` of `node.parent` is equal to that of `iparent`? I'm not sure what their exact relationship should be, but this `reg` property must definitely adhere to the `node.parent` one.

Reply:
I'll reread the specs more carefully.

Reply:
@57300 @bjarki-andreasen @nashif  @galak  It's looks like we have yet another DT interrupts related problem: `_interrupt_parent()` implementation.

The current implementation is scanning the tree for the 'interrupt-parent' until root node. But the latest DT specification is clear:
> Because the hierarchy of the nodes in the interrupt tree might not match the devicetree, the interrupt-parent property is available to make the definition of an interrupt parent explicit. The value is the phandle to the interrupt parent. **If this property is missing from a device, its interrupt parent is assumed to be its devicetree parent.**

And the latest Linux kernel already follow the above spec. Check the [vexpress-v2m.dtsi](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/arm/vexpress-v2m.dtsi), as an example. Obviously I should modify the `_interrupt_parent()` in this PR too, because the correct `interrupt-map` functionality  is depends on it, but consequences are not clear for me now. Since the `interrupt-map` property is actually used only for one board, I can make almost any modification for it. However, `_interrupt_parent()` function is an another beast: it's part of the core DT functionality, so a simple modification can break many targets. Should I put 'Architecture Review' on it (it's looks like "treewide" modif now)?

Reply:
To clarify, the issue is that we are walking up the tree until finding an interrupt controller, rather than asserting that the parent node is the interrupt controller, and generating a fault if this is not the case?

Reply:
> To clarify, the issue is that we are walking up the tree until finding an interrupt controller, rather than asserting that the parent node is the interrupt controller, and generating a fault if this is not the case?

A little bit different: the `_interrupt_parent()` should scan only and only the parent node, and use the `interrupt-parent` property, if it there, or use the parent node, if not. No matter if the parent node have or have not `interrupt-controller` property. In this case the ICU can have `interrupt-controller` (as it's more logical), and the `soc` become `interrupt-map` node (also using the address-cell and size from the `soc` becomes more agreeable). Check the 'vexpress-v2m.dtsi' file mentioned above, where interrupt-map is used instead of `interrupt-parent` (which would be clearer), but not instead of `interrupt-controller`.
However this new behavior of  `_interrupt_parent()` (and trivial modification), can break multiple targets. Or not. I haven't checked yet.

Reply:
I think we should be careful here. Firstly, the part of the spec that @avolkov-1221 quoted isn't new; git blame shows that the wording hasn't changed in 8 years: https://github.com/devicetree-org/devicetree-specification/blame/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter2-devicetree-basics.rst#L1076-L1080

But setting aside the spec for a moment, we should also take a look at the reference implementation - dtc, which is more actively maintained than the spec, as far as I'm aware. dtc will produce warnings on incorrect use of interrupt properties, and one of the checks is written here: https://github.com/dgibson/dtc/blob/99031e3a4a6e479466ae795790b44727434ca27d/checks.c#L1686

Here we can see that finding an interrupt parent does involve walking up the tree. The pseudocode would go something like this (written recursively):
```
proc get_interrupt_parent(node):
    if "interrupt-parent" in node.props:
        phandle := node.props["interrupt-parent"]
        iparent := get_node(phandle)
        assert iparent is interrupt controller or nexus
        return iparent

    if node is root:
        error "no interrupt parent"

    if node.parent is interrupt controller or nexus:
        return node.parent

    return get_interrupt_parent(node.parent)
```

This would suggest that the edtlib implementation is partially correct, only missing two things:
1. If an ancestor has an `interrupt-controller` or `interrupt-map` property, the walk must terminate.
2. If an `interrupt-parent` property is found, the linked node must be a valid interrupt controller or nexus.

I think it's strange that the spec seems to contradict this, by suggesting that only a node's immediate parent - not just any ancestor - can determine its interrupt parent. Even `vexpress-v2m.dtsi` doesn't abide by that rule - see how every interrupt-generating node (e.g. `/bus@40000000/motherboard-bus@40000000/iofpga@7,00000000/clcd@1f000`) is more than one level below the interrupt nexus (`/bus@40000000`).

On the other hand, I've found the spec to be incomplete and imprecise in several spots before. In fact, this very PR fixes a bug which likely stems from the spec's `interrupt-map` definition being misleading:
> **child unit address**
> The unit address of the child node being mapped. The number of 32-bit cells required to specify this is described by the *#address-cells* property of the **bus node on which the child is located**.

In conclusion, I think the best choice would be to align edtlib with dtc, which would also minimize the breakage. Regardless, I also think this can be taken in a separate PR, as the `#address-cells` fix is valuable enough on its own.

Reply:
nice research! @57300 :)

Reply:
> I think we should be careful here. Firstly, the part of the spec that @avolkov-1221 quoted isn't new; git blame shows that the wording hasn't changed in 8 years: https://github.com/devicetree-org/devicetree-specification/blame/2697c66a59b9177f8ff7813c355cc7aa05f6e40a/source/chapter2-devicetree-basics.rst#L1076-L1080

The 'new' here means newer than this comment in `_interrupt_parent()` :) :
```
 # Returns the node pointed at by the closest 'interrupt-parent', searching
 # the parents of 'node'. As of writing, this behavior isn't specified in
 # the DT spec., but seems to match what some .dts files except.
```
As you pointed out above, the behavior is specified in the spec for years, or the commenter meant something else.

>  
>  
... Snip ...

>This would suggest that the edtlib implementation is partially correct, only missing two things:
>
>If an ancestor has an interrupt-controller or interrupt-map property, the walk must terminate.
>If an interrupt-parent property is found, the linked node must be a valid interrupt controller or nexus.
>I think it's strange that the spec seems to contradict this, by suggesting that only a node's immediate parent - not just any ancestor - can determine its interrupt parent. Even vexpress-v2m.dtsi doesn't abide by that rule - see how every interrupt-generating node (e.g. /bus@40000000/motherboard-bus@40000000/iofpga@7,00000000/clcd@1f000) is more than one level below the interrupt nexus (/bus@40000000).

I suppose authors of the specification is mainly concerned with defining software behavior and seems to consider error checking outside its scope. Btw the `interrupt-parent` is propagated to all its children, isn't it? Also, dtc treats the `interrupt-map` the same way as `interrupt-controller` and `interrupt-parent`, i.e. as an inheritable property (I don't know where this behavior comes from).

... Snip ...

> 
> In conclusion, I think the best choice would be to align edtlib with dtc, which would also minimize the breakage. Regardless, I also think this can be taken in a separate PR, as the `#address-cells` fix is valuable enough on its own.

Agree that align `_interrupt_parent()` with dtc is the best solution for now, however I'm not sure if it would be reasonable to make this as a separate PR rather than a different commit in this one (actually it's simple, a "couple lines" modification). After updating `_raw_unit_addr()` et al, creating the nexus as done in vexpress-v2m.dtsi will still be impossible. I mean this part of the code:
```
	bus@40000000 {
		compatible = "simple-bus";
		#address-cells = <1>;
		#size-cells = <1>;
		....
		#interrupt-cells = <1>;
		interrupt-map-mask = <0 63>;
		interrupt-map = <0  0 &gic GIC_SPI  0 IRQ_TYPE_LEVEL_HIGH>,
		....
```
The current `_interrupt_parent()` skips such declarations (regardless of whether the `interrupt-parent` property is present or missing). Additionally, the intermediate interrupt controller will still need to be defined somewhat out of the spec: without `interrupt-controller`, but with the `interrupt-map` property. Probably it will be better to add yet another issue with ref to this PR.